### PR TITLE
Show/hide columns on propagate samples page

### PR DESCRIPTION
**Pull request by frankie13**
Don't forget, you need new things in the properties file for this to work!

### CODE REVIEW

**Comment by morgantaschuk** [comment]
If building with Java 8, something needs to be changed in the POM.

File: miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java [hunk 1]
```diff
@@ -859,16 +884,61 @@ public ModelAndView editPropagateSamples(@PathVariable String sampleIds, ModelMa
...
+          List<String> alias2Nums = new ArrayList<>(Arrays.asList(alias2.split(regex)));
+
+          alias1Nums.removeAll(Arrays.asList(""));
+          alias2Nums.removeAll(Arrays.asList(""));
+          if (alias1Nums.size() != alias2Nums.size()) {
+            // give up, just compare by default.
+            rtn = alias1.compareTo(alias2);
+          }
+          for (int i = 0; i < alias1Nums.size(); i++) {
+            if (Integer.valueOf(alias1Nums.get(i)) > Integer.valueOf(alias2Nums.get(i))) {
```
Comment (apmasell) [inline miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java]: If you prefer:
```
for (int i = 0; i < alias1Nums.size() && rtn == 0; i++) {
￼ rtn = Integer.compare(Integer.valueOf(alias1Nums.get(i)), Integer.valueOf(alias2Nums.get(i)));
}
```

File: core/src/main/java/uk/ac/bbsrc/tgac/miso/core/service/naming/DefaultNamingScheme.java [hunk 2]
```diff
@@ -47,7 +57,7 @@ public void setSampleAliasValidator(NameValidator validator) {
 
   @Override
   public void setLibraryAliasGenerator(NameGenerator<Library> generator) {
-    this.libraryAliasGenerator = generator;
+    this.libraryAliasGenerator = (DefaultLibraryAliasGenerator) generator;
```
Comment (djcooke) [inline core/src/main/java/uk/ac/bbsrc/tgac/miso/core/service/naming/DefaultNamingScheme.java]: Remove cast

File: core/src/main/java/uk/ac/bbsrc/tgac/miso/core/service/naming/DefaultNamingScheme.java [hunk 3]
```diff
@@ -22,9 +24,17 @@
   private NameValidator sampleAliasValidator = new DefaultSampleAliasValidator();
   private NameGenerator<Sample> sampleAliasGenerator = null;
   private NameValidator libraryAliasValidator = new DefaultLibraryAliasValidator();
-  private NameGenerator<Library> libraryAliasGenerator = new DefaultLibraryAliasGenerator();
+  private DefaultLibraryAliasGenerator libraryAliasGenerator = new DefaultLibraryAliasGenerator();
```
Comment (djcooke) [inline core/src/main/java/uk/ac/bbsrc/tgac/miso/core/service/naming/DefaultNamingScheme.java]: `NameGenerator` must remain the declared type - configuration may cause this to be replaced by a different generator

File: miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java [hunk 4]
```diff
@@ -845,6 +866,10 @@ public ModelAndView editPropagateSamples(@PathVariable String sampleIds, ModelMa
           hasPlain = true;
         }
         LibraryDto library = new LibraryDto();
+        Library tempLibrary = new LibraryImpl();
+        tempLibrary.setSample(sample);
+        final String libraryAlias = namingScheme.generateLibraryAlias(tempLibrary);
```
Comment (djcooke) [inline miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java]: Alias generation should be handled in the Service layer unless it's necessary to display the alias before the library is saved?

Comment (frankie13) [inline miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java]: it is I'm afraid.  Without an alias the form is invalid so we need to know it upfront.

**Comment by frankie13** [comment]
@djcooke @morgantaschuk I believe I have addressed all the things!

File: miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java (hunk 4 above)
Comment (djcooke) [inline miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java]: That should be changed to match how the sample forms work: If sample alias generation is enabled, it is not a required field. It may still be provided if you'd like to use a custom alias, but leaving it blank will result in one being generated.

Generating the alias during save is also better for generating unique names. If it's done during creation and two people are creating libraries at the same time, it can result in a collision.

File: miso-web/src/main/webapp/pages/bulkEditLibraries.jsp [hunk 5]
```diff
@@ -78,15 +78,15 @@
...
- 
+
  </div>
-  
+
   <script type="text/javascript">
     jQuery(document).ready(function () {
       Library.hot.librariesJSON = ${librariesJSON};
-      Library.hot.librariesJSON = Hot.sortByProperty(Library.hot.librariesJSON, 'parentSampleAlias');
+      Library.hot.hideCols = ${hideCols};
```
Comment (callunity) [inline miso-web/src/main/webapp/pages/bulkEditLibraries.jsp]: `${hideCols}` isn't set for bulk editing, which causes critical JS errors.

File: miso-web/src/main/webapp/pages/bulkEditLibraries.jsp [hunk 6]
```diff
@@ -78,15 +78,15 @@
...
    <div id="hotContainer"></div>
- 
+
  </div>
-  
+
   <script type="text/javascript">
     jQuery(document).ready(function () {
       Library.hot.librariesJSON = ${librariesJSON};
-      Library.hot.librariesJSON = Hot.sortByProperty(Library.hot.librariesJSON, 'parentSampleAlias');
```
Comment (callunity) [inline miso-web/src/main/webapp/pages/bulkEditLibraries.jsp]: Please wrap the `Hot.sortByProperty` line if an `if Hot.detailedSample` check (and move the `Hot.detailedSample` assignment up above this line) instead of deleting it, as we will still want to use this sorting in our lab.

Comment (frankie13) [inline miso-web/src/main/webapp/pages/bulkEditLibraries.jsp]: :white_check_mark:

File: miso-web/src/main/webapp/pages/bulkEditLibraries.jsp (hunk 5 above)
Comment (frankie13) [inline miso-web/src/main/webapp/pages/bulkEditLibraries.jsp]: :white_check_mark:

File: miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java (hunk 4 above)
Comment (frankie13) [inline miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java]: I agree with you @djcooke - this approach would be cleaner.
We do however have something that works, and works well for now and given that this task has already been dragging on for sometime I suggest we go with this code as it stands with a view to changing the approach in the future.
I am concerned with getting the develop branch deployed to our production server soon as I had hoped to do it around new year, since this issue isn't blocking / breaking anything I recommend we press on!

File: miso-web/src/main/webapp/pages/bulkEditLibraries.jsp [hunk 7]
```diff
@@ -78,20 +78,32 @@
...
+
+      if (Hot.detailedSample) {
+        Library.hot.librariesJSON = Hot.sortByProperty(Library.hot.librariesJSON, 'parentSampleAlias');
+      }
+      <c:choose>
+	    <c:when test="${not empty hideCols}">
+	      Library.hot.hideCols = ${hideCols};
+	    </c:when>    
+	    <c:otherwise>
+	      Library.hot.hideCols = {};
```
Comment (callunity) [inline miso-web/src/main/webapp/pages/bulkEditLibraries.jsp]: Since you're initializing `${hideCols}` as a JSONArray, you could also just make the assignment instead of wrapping it in the test; if you keep it this way, I think this needs to be an empty array instead of empty object since you're doing checks for `indexOf`.

File: miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java [hunk 8]
```diff
@@ -775,6 +793,10 @@ public ModelAndView editPropagateSamples(@PathVariable String sampleIds, ModelMa
           hasPlain = true;
         }
         LibraryDto library = isDetailedSampleEnabled() ? new DetailedLibraryDto() : new LibraryDto();
+        Library tempLibrary = new LibraryImpl();
+        tempLibrary.setSample(sample);
+        final String libraryAlias = namingScheme.generateLibraryAlias(tempLibrary);
```
Comment (callunity) [inline miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java]: Please wrap 798 and 799 in `if (namingScheme.hasLibraryAliasGenerator())`

File: miso-web/src/main/resources/external/miso.properties [hunk 9]
```diff
@@ -39,6 +39,10 @@ miso.alerting.enabled:true
 # toggle mirroring of all alerts to system alerts (user id 0)
 miso.alerting.saveSystemAlerts:true
 
+## configs for caching
+# precache objects at MISO startup. This will greatly increase startup time, but will improve MISO performance for users at the outset
+miso.db.caching.precache.enabled:false
```
Comment (callunity) [inline miso-web/src/main/resources/external/miso.properties]: This won't do much since Hibernate takes care of the caching, but that means it can be deleted later :)
